libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,567 @@
+	fixStackTooDeep(_entry);
+}
+
+Stack StackLayoutGenerator::combineStack(Stack const& _stack1, Stack const& _stack2)
+{
+	if (_stack1.empty())
+		return _stack2;
+	if (_stack2.empty())
+		return _stack1;
+
+	// TODO: there is probably a better way than brute-forcing. This has n! complexity or worse, so
+	// we can't keep it like this.
+
+	Stack commonPrefix;
+	for (auto&& [slot1, slot2]: ranges::zip_view(_stack1, _stack2))
+	{
+		if (!(slot1 == slot2))
+			break;
+		commonPrefix.emplace_back(slot1);
+	}
+	Stack stack1Tail = _stack1 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+	Stack stack2Tail = _stack2 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+
+	Stack candidate;
+	for (auto slot: stack1Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	for (auto slot: stack2Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	cxx20::erase_if(candidate, [](StackSlot const& slot) {
+		return holds_alternative<LiteralSlot>(slot) || holds_alternative<FunctionCallReturnLabelSlot>(slot);
+	});
+
+	std::map<size_t, Stack> sortedCandidates;
+
+	// TODO: surprisingly this works for rather comparably large candidate size, but we should probably
+	// set up some limit, since this will quickly explode otherwise.
+	// Ideally we would then have a better fallback mechanism - although returning any naive union of both stacks
+	// like ``candidate`` itself may just be fine.
+	//	if (candidate.size() > 8)
+	//		return candidate;
+
+	auto evaluate = [&](Stack const& _candidate) -> size_t {
+		size_t numOps = 0;
+		Stack testStack = _candidate;
+		auto swap = [&](unsigned _swapDepth) { ++numOps; if (_swapDepth > 16) numOps += 1000; };
+		auto dup = [&](unsigned _dupDepth) { ++numOps; if (_dupDepth > 16) numOps += 1000; };
+		auto push = [&](StackSlot const& _slot) {
+			if (!canBeFreelyGenerated(_slot))
+			{
+				auto offsetInPrefix = util::findOffset(commonPrefix, _slot);
+				yulAssert(offsetInPrefix, "");
+				// Effectively this is a dup.
+				++numOps;
+				// TODO: Verify that this is correct. The idea is to penalize dupping stuff up that's too deep in
+				//       the prefix at this point.
+				if (commonPrefix.size() + testStack.size() - *offsetInPrefix > 16)
+					numOps += 1000;
+			}
+		};
+		createStackLayout(testStack, stack1Tail, swap, dup, push, [&](){} );
+		testStack = _candidate;
+		createStackLayout(testStack, stack2Tail, swap, dup, push, [&](){});
+		return numOps;
+	};
+
+	// See https://en.wikipedia.org/wiki/Heap's_algorithm
+	size_t n = candidate.size();
+	sortedCandidates.insert(std::make_pair(evaluate(candidate), candidate));
+	std::vector<size_t> c(n, 0);
+	size_t i = 1;
+	while (i < n)
+	{
+		if (c[i] < i)
+		{
+			if (i & 1)
+				std::swap(candidate.front(), candidate[i]);
+			else
+				std::swap(candidate[c[i]], candidate[i]);
+			sortedCandidates.insert(std::make_pair(evaluate(candidate), candidate));
+			++c[i];
+			++i;
```

Comment:
```suggestion
			i = 0;
```

According to the procedural algorithm in https://en.wikipedia.org/wiki/Heap's_algorithm.

Kinda interesting that this still leads to better stack?

Reply:
Actually, changing to `i = 0`, makes isoltest extremely slow :( Not sure if you left `i = 0` there just for testing.

Reply:
Ah, nice! I was quite puzzled why this wasn't running much slower actually, but never rechecked this part again :-).
It's quite natural that it makes runs extremely slow, if we do this in all cases - it's exponential after all, but we can just cut it short and go with a "naive merge" for larger stacks.
I think I mentioned it elsewhere and said we could consider something like ``if (un)likely`` markers - the main thing this does is to move the required shuffling between control flow paths, I don't think going out of our way to choose a particular optimal one is really worth it here.

Ideally, of course, we'd find a non-exponential way to do this in general, if only as approximation.

Reply:
Playing around with it a bit, it *is* indeed interesting that changing this and instead just using the initial layout for large candidates does indeed have a noticable adverse effect - maybe worth thinking about which subset of permutations this error actually led to :-).
It's somewhat surprising that even keeping the erroneous algorithm for large candidates and using the proper one for smaller ones can make things worse - but that's probably due to the layouts we choose here having global effects, so it's impossible to know which one is absolutely best anyways, since reducing swaps at some control flow join may increase swaps in a different one which may be executed more often.

Reply:
Actually, we could even mark some conditional jumps as "likely" during the control flow graph generation - for example it's fair to assume that in a loop the path with the condition being ``true`` is executed more often and worth preferring. We can also try taking reverting control flow into account - a lot of options to tweak this and it will always be a heuristics in the end.

Reply:
In any case, this was always rather intended as placeholder and there's room to play around with it - I'd actually be rather curious about what's the theoretical minimal complexity for an algorithm that solves this optimally...